Support for albums

lyrra opened the pull request:
Draft, TBD before PR:
- Would be good to have this tested by others.
- Also it'd be great to get @SKefalidis' official blessing and review

Current bugs (to be fixed):
- [ ] crash if closing score
- [x] crash if closing album (in album-manager) FIXED in f184fa8a
- [ ] crash if closing album (in album-manager) IF two identical scores are in album
- [x] crash if "save album" FIXED in 118b96c3
- [x] ~~crash if exiting~~ (seems not related to this PR)
- [ ] Writes album properties in an non-album mscx file

---


(squashed all of Sergios - Anestis Kefalidis patches (from #6145), and then rebased onto 3.x)
~~(it compiles almost on mingw, (get undefined reference to `__stack_chk_fail', something with -fstack-protector))~~
Compiles in a clean build.

---

fixed currentScoreView updated correctly when switching between excerpts

fixed crash on close album

possible fix-workaround for session restore related crashes

fix paint crash

fix albummanagerdialog with no album loaded

fixed albummanager window calls hide event when changing workspace

drawFrontCover

album title loaded

modified PR #6352

contents pages

fixed crash when adding scores with different instrumentation on new album + playback fix

read + collect_artifacts

enable albumBreaks test

default pause and mass change pause

fixed problems related to restore session? (finally?)

codestyle

use albumScores() instead of x->score

tests for legacy albums

simplified and improved playback control

fixed seq crash

fixed restore session

export parts

at last good playback!!!

fixed crashes related to deleted Album pages

codestyle

cleanup + small things

fix build failure + codestyle

forgot this file

fixed mtests

fixed mtest failure

readme

more cleanup and shorter mtest-scores

forgot this again

correct undo

fixed various issues

updateContents when disabling, crash fixes, drawFrontCover in .msca

disabled Album Manager playback

updated tests

crash fixes

fixed warning + less doLayout calls

fixed parts editing + shadownote for parts

comments

codestyle

rename and comments

a few got left behind

codestyle

rename, delete old code, simplify

codestyle

a bunch of contents and renames

forgot a couple of things

updated text and readme

changed how scores are deleted and forgot this

fix crash by resetting tempScoreTabindex when closing an album

another small fix

fix crash when drag&drop with disabled movements, added support for enabling/disabling for parts, various other improvements

readme

removed text

fixed editing parts and updating the view when editing them

parts scoreview constraints

final touches

readme

codestyle

fix travis failure

FIRST COMMIT AFTER GOOGLE SUMMER OF CODE:\
Added tabstops and cleaned ui file, again

text styles for front cover and contents wip

text styles for front cover and contents

fixed rebase

removed unused variable

codestyle

fixed build failure

preferences

SKefalidis committed:
squash for appveyor

fixed currentScoreView updated correctly when switching between excerpts

fixed crash on close album

possible fix-workaround for session restore related crashes

fix paint crash

fix albummanagerdialog with no album loaded

fixed albummanager window calls hide event when changing workspace

drawFrontCover

album title loaded

modified PR #6352

contents pages

fixed crash when adding scores with different instrumentation on new album + playback fix

read + collect_artifacts

enable albumBreaks test

default pause and mass change pause

fixed problems related to restore session? (finally?)

codestyle

use albumScores() instead of x->score

tests for legacy albums

simplified and improved playback control

fixed seq crash

fixed restore session

export parts

at last good playback!!!

fixed crashes related to deleted Album pages

codestyle

cleanup + small things

fix build failure + codestyle

forgot this file

fixed mtests

fixed mtest failure

readme

more cleanup and shorter mtest-scores

forgot this again

correct undo

fixed various issues

updateContents when disabling, crash fixes, drawFrontCover in .msca

disabled Album Manager playback

updated tests

crash fixes

fixed warning + less doLayout calls

fixed parts editing + shadownote for parts

comments

codestyle

rename and comments

a few got left behind

codestyle

rename, delete old code, simplify

codestyle

a bunch of contents and renames

forgot a couple of things

updated text and readme

changed how scores are deleted and forgot this

fix crash by resetting tempScoreTabindex when closing an album

another small fix

fix crash when drag&drop with disabled movements, added support for enabling/disabling for parts, various other improvements

readme

removed text

fixed editing parts and updating the view when editing them

parts scoreview constraints

final touches

readme

codestyle

fix travis failure

FIRST COMMIT AFTER GOOGLE SUMMER OF CODE:\
Added tabstops and cleaned ui file, again

text styles for front cover and contents wip

text styles for front cover and contents

fixed rebase

removed unused variable

codestyle

fixed build failure

preferences

lyrra commented:
- crash if closing score
- crash if closing album

Jojo-Schmitz commented:
Seems a rebase is needed

Jojo-Schmitz commented:
And chance to get those crashes resolved?

Jojo-Schmitz commented:
Also the mtests are failing

lyrra committed:
bugfix, check album is set (isAlbum)

lyrra committed:
albums bugfix: remove dead pages

lyrra committed:
albums: mtest musicxml default album-font-sizes

lyrra committed:
albums: code-formatting

Jojo-Schmitz commented on mtest/musicxml/io/testAdditionalFermatas_ref.mscx:
```diff
@@ -34,6 +34,11 @@
+      <albumFrontComposerFontSize>10</albumFrontComposerFontSize>
+      <albumFrontLyricistFontSize>10</albumFrontLyricistFontSize>
+      <albumContentsTitleFontSize>10</albumContentsTitleFontSize>
+      <albumContentsTextFontSize>10</albumContentsTextFontSize>
```
Not sure this is the right fix, default values are not written to score IIRC, so here the default might be wrong?

lyrra replied:
Perhaps those are written unconditionally to the score-file, even if albums feature are not used. This breaks file-version since 3.6.3 might not be able to load an 3.7+Albums score-file.

lyrra replied:
It seems musicxml during import creates all fields unconditionally, resulting in an blown-up mscx.

Jojo-Schmitz replied:
True, I see other settings there too, strangely all with a font size of 10pt.

lyrra commented on mscore/albummanager.h:
```diff
@@ -13,39 +13,118 @@
 #ifndef __ALBUMMANAGER_H__
 #define __ALBUMMANAGER_H__
 
+#include <optional>
```
macos < 10.14 doesn't have full c++17 support, need to back out optional use.
See https://github.com/octobanana/peaclock/issues/2

lyrra committed:
bugfix: ensure currentScore->systems is not empty

lyrra committed:
albums: mtest update musescore-version

lyrra committed:
UNVERIFIED albums: mtest updates

not sure if these changes to the reference-scores are
valid, or if the albums-changes are wrongly causing them.

lyrra committed:
albums: bugfix null-check m_topMovement

Jojo-Schmitz commented:
Would be good tohave this tested by others

Jojo-Schmitz commented:
Also it's be great to get @SKefalidis' official blessing and review

SKefalidis commented:
Hi @Jojo-Schmitz (long time no see, hope you are doing well :-) and @lyrra what are you planning to do with this PR? Are you guys working on a MuseScore 3 fork?

Jojo-Schmitz commented:
Yes, a 3.x fork. I'm working on it since quite long now.

Jojo-Schmitz commented:
There's yet another merge conflict, mind to rebase?

lyrra commented:
> There's yet another merge conflict (an easy one IMHO), mind to rebase?

Thanks for the heads-up! Rebased.

lyrra commented:
"@[Jojo-Schmitz](https://github.com/Jojo-Schmitz) Jojo-Schmitz deleted the branch Jojo-Schmitz:3.x 1 hour ago"
No more 3.x?  😱 🤯😭 😉

Jojo-Schmitz commented:
2 mtests fail now, apparently on some added tests, should be easy to fix, by adding the correspunding lines

Jojo-Schmitz commented:
I guess the real question is why the mtests want to see some non-default text sizes?
```
+      <albumFrontTitleFontSize>10</albumFrontTitleFontSize>
+      <albumFrontComposerFontSize>10</albumFrontComposerFontSize>
+      <albumFrontLyricistFontSize>10</albumFrontLyricistFontSize>
+      <albumContentsTitleFontSize>10</albumContentsTitleFontSize>
+      <albumContentsTextFontSize>10</albumContentsTextFontSize>
```

Jojo-Schmitz commented on mscore/musescore.cpp:
```diff
@@ -2233,7 +2270,7 @@ void MuseScore::updateMenus()
 #endif
       connect(openRecent,     SIGNAL(aboutToShow()),       SLOT(openRecentMenu()));
-      connect(openRecent,     SIGNAL(triggered(QAction*)), SLOT(selectScore(QAction*)));
+    connect(openRecent,     SIGNAL(triggered(QAction*)), SLOT(selectFile(QAction*)));
```
formatting

Jojo-Schmitz commented on mscore/file.cpp:
```diff
@@ -309,7 +313,9 @@ void MuseScore::openFiles(bool switchTab, bool singleFile)
-             << tr("MuseScore Backup Files") + " (*.mscz, *.mscx,)";
+           << tr("MuseScore Backup Files") + " (*.mscz, *.mscx,)"
+           << tr("MuseScore 4 Album Files") + " (*.msca *.mscaz)"
+           << tr("MuseScore 2 Album Files") + " (*.album)";
```
formatting

lyrra replied:
Also:
MuseScore **4** Album Files  -> MuseScore **3** Album Files

Jojo-Schmitz replied:
yes, of course

Jojo-Schmitz commented:
And now a merge conflict to resolve

lyrra committed:
indent fixes

lyrra committed:
music-xml font-sizes

(squash) with patch:
  * albums: mtest musicxml default album-font-sizes

lyrra commented:
> 2 mtests fail now, apparently on some added tests, should be easy to fix, by adding the corresponding lines
> 
> OTOH: why are those needed, if not deviating from the default? IIRC we don't write default values into mscx files

Looking at the surrounding values of other settings, it seems it is writing default values.
What is more worrisome is why we are writing albums key/vals at all in an non-album mscx file.

Jojo-Schmitz commented:
mtest failures now, crashes actually

lyrra commented:
> mtest failures now, crashes actually

Thanks for the heads-up! pushed fixes

Jojo-Schmitz commented:
Once again a rebase is needed

Jojo-Schmitz commented:
I believe a rebase would be good (and needed anyway)

Jojo-Schmitz commented:
@lyrra could you please rebase? The logs and artifacts have expired.

Jojo-Schmitz commented on mscore/albummanager.ui:
```diff
@@ -1,174 +1,242 @@
+           <string notr="true"/>
+          </property>
+          <property name="icon">
+           <iconset resource="../../../../../../../tmp/QtCreator-FLYAKj/qtc-cmake-rKuWjOTJ/mscore/musescore_qmlcache.qrc">
```
Huh?

Jojo-Schmitz commented on mscore/albummanager.ui:
```diff
@@ -1,174 +1,242 @@
+           <string notr="true"/>
+          </property>
+          <property name="icon">
+           <iconset resource="../../../../../../../tmp/QtCreator-FLYAKj/qtc-cmake-rKuWjOTJ/mscore/musescore_qmlcache.qrc">
```
same

Jojo-Schmitz commented:
The mtest reveal that these settings are (expected to be)
```xml
      <albumFrontTitleFontSize>10</albumFrontTitleFontSize>
      <albumFrontComposerFontSize>10</albumFrontComposerFontSize>
      <albumFrontLyricistFontSize>10</albumFrontLyricistFontSize>
      <albumContentsTitleFontSize>10</albumContentsTitleFontSize>
      <albumContentsTextFontSize>10</albumContentsTextFontSize>
```
in the mtest scores.
But as per style.cpp  they are 36, 16, 16,24 and 14, repectively. Question now is where that `10` stems from and why.
I don't see wherte and why that happens

Jojo-Schmitz commented:
Mind to rebase and fix the merge conflicts?

lyrra committed:
fix mtest

album fontsize is implicitly tucked on

Jojo-Schmitz commented:
Mind to rebase and fix the merge conflicts once again?

lyrra committed:
cant set title to non-existing album

lyrra committed:
album title edits dont access front cover VBox if null

lyrra committed:
cmake: fix mtest

lyrra committed:
fix rebase text.h
